algorithms/linfa-tsne/src/hyperparams.rs

```diff
@@ -0,0 +1,160 @@
+/// distributions with respect to the locations of the points in the map.
+///
+/// This crate wraps the [bhtsne](https://github.com/frjnn/bhtsne) crate for the linfa project. It
+/// implements the exact t-SNE, as well as the Barnes-Hut approximation.
+///
+/// # Examples
+///
+/// ```ignore
+/// use linfa::traits::Transformer;
+/// use linfa_tsne::TSne;
+///
+/// let ds = linfa_datasets::iris();
+///
+/// let ds = TSne::embedding_size(2)
+///     .perplexity(10.0)
+///     .approx_threshold(0.6)
+///     .transform(ds);
+/// ```
+
+/// A verified hyper-parameter set ready for prediction
+pub struct TSneValidParams<F, R> {
+    embedding_size: usize,
+    approx_threshold: F,
+    perplexity: F,
+    max_iter: usize,
+    preliminary_iter: Option<usize>,
+    rng: R,
+}
+
+impl<F, R> TSneValidParams<F, R> {
+    pub fn embedding_size(&self) -> usize {
+        self.embedding_size
+    }
+
+    pub fn approx_threshold(&self) -> F {
+        self.approx_threshold
+    }
+
+    pub fn perplexity(&self) -> F {
+        self.perplexity
+    }
+
+    pub fn max_iter(&self) -> usize {
+        self.max_iter
+    }
+
+    pub fn preliminary_iter(&self) -> &Option<usize> {
+        &self.preliminary_iter
+    }
+
+    pub fn rng(&self) -> &R {
+        &self.rng
+    }
+}
+
+pub struct TSneParams<F, R>(TSneValidParams<F, R>);
+
+impl<F: Float> TSneParams<F, SmallRng> {
+    /// Create a t-SNE param set with given embedding size
+    ///
+    /// # Defaults to:
+    ///  * `approx_threshold`: 0.5
+    ///  * `perplexity`: 5.0
+    ///  * `max_iter`: 2000
+    ///  * `rng`: SmallRng with seed 42
+    pub fn embedding_size(embedding_size: usize) -> TSneParams<F, SmallRng> {
+        Self::embedding_size_with_rng(embedding_size, SmallRng::seed_from_u64(42))
+    }
+}
+
+impl<F: Float, R: Rng + Clone> TSneParams<F, R> {
+    /// Create a t-SNE param set with given embedding size and random number generator
+    ///
+    /// # Defaults to:
+    ///  * `approx_threshold`: 0.5
+    ///  * `perplexity`: 5.0
+    ///  * `max_iter`: 2000
+    pub fn embedding_size_with_rng(embedding_size: usize, rng: R) -> TSneParams<F, R> {
+        Self(TSneValidParams {
+            embedding_size,
+            rng,
+            approx_threshold: F::cast(0.5),
+            perplexity: F::cast(5.0),
+            max_iter: 2000,
+            preliminary_iter: None,
+        })
+    }
+
+    /// Set the approximation threshold of the Barnes Hut algorithm
+    ///
+    /// The threshold decides whether a cluster centroid can be used as a summary for the whole
+    /// area. This was proposed by Barnes and Hut and compares the ratio of cell radius and
+    /// distance to a factor theta. This threshold lies in range (0, inf) where a value of 0
+    /// disables approximation and a positive value approximates the gradient with the cell center.
+    pub fn approx_threshold(mut self, threshold: F) -> Self {
+        self.0.approx_threshold = threshold;
+
```

Comment:
Don't think you should put empty lines here since the methods are all 2-liners anyways.